### PR TITLE
Allow to lookup users based on a custom attribute.

### PR DESCRIPTION
**Pull request by ticosax**
Useful for custom User models that define another unique id field for instance.

### CODE REVIEW

**Comment by jpadilla** [comment]
@ticosax working on something that is a bit related to this on https://github.com/GetBlimp/django-rest-framework-jwt/commit/48b2c086d71d485b75f100a83179324b7feb1ece and #101.

**Comment by ticosax** [comment]
Thanks I will give it a try

**Comment by ticosax** [comment]
@jpadilla I tested you current dev branch.
While It technically works well, it doesn't serve my needs.
Let me explain in detail:
The field I'm relying to perform user lookup is a `UUIDField`.
Following the intentation of the patch in dev branch, I define `USERNAME__FIELD = 'uuid'` on my User model.
 It means that from now on, every django app that will consume the `User` model will use `uuid` instead of `username`. The problem is that it doesn't work quite well with all the third party libraries I'm using (`python-social-auth` for instance),  because most of the time, `username` field is assumed to be a `CharField`. It means some libraries will try to create a `User` by providing `username` that is not suitable for `UUIDField`.

To recap, while it is working fine with drf-jwt, it doesn't play well with other third party libraries.

**Comment by jpadilla** [comment]
After reading this again I think there's still room for improvement here, probably by introducing an overridable function/setting that can change the current behavior in `VerificationBaseSerializer._check_user` and `BaseJSONWebTokenAuthentication. authenticate_credentials`.
